doc/languages-frameworks/maven.md

```diff
@@ -0,0 +1,351 @@
+  buildPhase = ''
+    mvn package -Dmaven.repo.local=$out
+  '';
+
+  # keep only *.{pom,jar,sha1,nbm} and delete all ephemeral files with lastModified timestamps inside
+  installPhase = ''
+    find $out -type f \
+      -name \*.lastUpdated -or \
+      -name resolver-status.properties -or \
+      -name _remote.repositories \
+      -delete
+  '';
+
+  # don't do any fixup
+  dontFixup = true;
+  outputHashAlgo = "sha256";
+  outputHashMode = "recursive";
+  # replace this with the correct SHA256
+  outputHash = stdenv.lib.fakeSha256;
+}
+```
+
+The build will fail, and tell you the expected `outputHash` to place. When you've set the hash, the build will return with a _/nix/store_ entry whose contents are the full Maven repository.
+
+> Some additional files are deleted that would cause the output hash to change potentially on subsequent runs.
+
+```bash
+❯ tree $(nix-build --no-out-link double-invocation-repository.nix) | head
+/nix/store/8kicxzp98j68xyi9gl6jda67hp3c54fq-maven-repository
+├── backport-util-concurrent
+│   └── backport-util-concurrent
+│       └── 3.1
+│           ├── backport-util-concurrent-3.1.pom
+│           └── backport-util-concurrent-3.1.pom.sha1
+├── classworlds
+│   └── classworlds
+│       ├── 1.1
+│       │   ├── classworlds-1.1.jar
+```
+
+If your package uses _SNAPSHOT_ dependencies or _version ranges_; there is a strong likelihood that over-time your output hash will change since the resolved dependencies may change.
+
+## Building a JAR
+
+Regardless of which strategy is chosen above, the step to build the derivation is the same.
+
+```nix
+{ pkgs ? import <nixpkgs> { }, stdenv ? pkgs.stdenv, lib ? pkgs.lib
+, maven ? pkgs.maven, callPackage ? pkgs.callPackage }:
+# pick a repository derivation, here we will use buildMaven
+let repository = callPackage ./build-maven-repository.nix { };
+in stdenv.mkDerivation rec {
+  pname = "maven-demo";
+  version = "1.0";
+
+  src = builtins.fetchTarball "https://github.com/fzakaria/nixos-maven-example/archive/main.tar.gz";
+  buildInputs = [ maven ];
+
+  buildPhase = ''
+    echo "Using repository ${repository}"
+    mvn --offline -Dmaven.repo.local=${repository} package;
+  '';
+
+  installPhase = ''
+    install -Dm644 target/${pname}-${version}.jar $out/share/java
+  '';
+}
+```
+
+> We place the library in _$out/share/java_ since JDK package has a _stdenv setup hook_ that adds any JARs in the _share/java_ directories of the build inputs to the CLASSPATH environment.
+
+```bash
+❯ tree $(nix-build --no-out-link build-jar.nix)
+/nix/store/7jw3xdfagkc2vw8wrsdv68qpsnrxgvky-maven-demo-1.0
+└── share
+    └── java
+        └── maven-demo-1.0.jar
+
+2 directories, 1 file
+```
+
+## Runnable JAR
+
+You are able to run `mvn compile` but how do you get a runnable artifact without having to use something like building an UberJar ?
+
+We will use the same repository we built above (either _double invocation_ or _buildMaven_) to setup a CLASSPATH for our JAR.
+
+### CLASSPATH
```

Comment:
I'm a little bit confused about this section. In the derivation of it you create a wrapper, and that calls Main as it seems. What's the difference between this and the last example expression?

Reply:
So fundamentally, your Java invocation needs to have **all** dependencies on the CLASSPATH.
> CLASSPATH is the search path Java uses to lookup classes

The CLASSPATH can be set explicitly within a JAR which the subsequent section does through the help of Maven via the plugin via the MANIFEST file.

Here is the contents of the JAR and we see the CLASSPATH is set and expected at a fixed location relative to the JAR.
```bash
❯ unzip -q -c /nix/store/8d4c3ibw8ynsn01ibhyqmc1zhzz75s26-maven-demo-1.0/share/java/maven-demo-1.0.jar
Manifest-Version: 1.0
Archiver-Version: Plexus Archiver
Built-By: nixbld
Class-Path: . ../../repository/com/vdurmont/emoji-java/5.1.1/emoji-jav
 a-5.1.1.jar ../../repository/org/json/json/20170516/json-20170516.jar
Created-By: Apache Maven 3.6.3
Build-Jdk: 1.8.0_265
Main-Class: Main
```

The first approach though instead builds the JAR normally and then augments the _makeWrapper_ to set the CLASSPATH via the command line.

You've given feedback on providing context on which is appropriate which I will do.

Fundamentally, if you are a _nixpkgs_ author wrapping another project, the method by setting the CLASSPATH via _makeWrapper_ is preferable since it's a bit more of a pain to create a patch file to change the project's pom.xml

If you are a project writer and using nix-shell or nix-build to build your project, it would be easier for you to modify your pom to add the required plugin.

Reply:
I don't read your comments, as I expect these explanations to be in the documentation.